src/content/docs/references/docs/optionals_advanced.md

```diff
@@ -0,0 +1,355 @@
+---
+title: Optionals - Advanced
+description: Optionals - Advanced
+sidebar:
+    order: 115
+---
+
+These features are *nice to haves*, you should be able to do what you need mostly using the [standard Optionals features](./optionals_intro.md).
+
+### Optionals are only defined in certain code
+✅ Variable declarations
+```c3
+int! example = unreliable_function();
+```
+✅ Function return signature
+```c3
+fn int! example() { ... }
+```
+
+## Handling `Empty` Optional
+
+
+### File reading example 
+Let's expand this example to something which *may* fail, like opening a file. 
+
+- If the file is present the `Result` will be the first 100 bytes of the file.
+- If the file is not present the `Excuse` will be `IoError.FILE_NOT_FOUND`.
+
+Try running this code below with and without a file called `file_to_open.txt` in the same directory.
+
+```c3
+import std::io;
+
+/** 
+ * Function modifies `buffer`
+ * Returns Optional with `Result` of type `char[]` or an `Excuse`
+**/
+fn char[]! read_file(String filename, char[] buffer)
+{
+    // Return `Excuse` if failed opening file, using rethrow `!`
+    File! file = file::open(filename, "r")!; 
+
+    // At scope exit, close the file if it was opened successfully
+    // Discard the `Excuse` from file.close() with (void) cast
+    defer (void)file.close(); 
+
+    // Return `Excuse` if failed to read file, using rethrow `!`
+    file.read(buffer)!; 
+    return buffer; // return the buffer `Result`
+}
+
+fn void! main()
+{
+    char[] buffer = mem::new_array(char, 100);
+    defer free(buffer); // Free memory on scope exit
+
+    // Catch missing `Result` with an underlying `Excuse` defined, assign to `excuse`
+    char[]! read_buffer = read_file("file_to_open.txt", buffer);
+    if (catch excuse = read_buffer) 
+    {
+        io::printfn("Excuse found: %s", excuse);
+        return excuse?; // Returning `Excuse` using `?` suffix
+    }
+
+    // `read_buffer` behaves like a normal variable here 
+    // because missing `Result` was handled by scope exit in `if (catch)`
+    io::printfn("read_buffer: %s", read_buffer);
+}
+```
+
+### Using force unwrap `!!` to panic on `Empty` Optional
+
+The force unwrap `!!` will issue a panic and exit the program if Optional is `Empty`.
```

Comment:
Add something like "this is useful when the error should – in normal cases – not happen and you don't want to write any error handling for it. That said, it should be used with caution in production code"